cdap-common/src/test/java/co/cask/cdap/io/SchemaTest.java

```diff
@@ -373,6 +373,98 @@ public void testSerializable() throws IOException, ClassNotFoundException {
     }
   }
 
+  /**
+   * Parent class for testing recursive schema.
+   */
+  private static final class NodeParent {
+    private List<NodeChild> children;
+  }
+
+  /**
+   * Child class for testing recursive schema.
+   */
+  private static final class NodeChild {
+    private int data;
+    private NodeParent parent;
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  @Test
+  public void testGeneratorRecursion() throws Exception {
+    // This test the schema generator be able to handle recursive schema between multiple classes
+    Schema generatedSchema = new ReflectionSchemaGenerator(false).generate(NodeParent.class);
+
+    // Validate the NodeChild.parent schema is the same as the schema of the Parent class
+    // The schema generator always generate array value as nullable union
+    Schema childSchema = generatedSchema.getField("children").getSchema().getComponentSchema().getNonNullable();
+    Assert.assertEquals(generatedSchema, childSchema.getField("parent").getSchema());
+
+    // Serialize the schema and then parse it back
+    Schema deserializedSchema = Schema.parseJson(generatedSchema.toString());
+    Assert.assertEquals(generatedSchema, deserializedSchema);
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  @Test
+  public void testRecursion() throws Exception {
+    // Test simple recursion
+    Schema nodeSchema = Schema.recordOf("Node",
+                                        Schema.Field.of("data", Schema.of(Schema.Type.INT)),
+                                        Schema.Field.of("next", Schema.recordOf("Node")));
+    Assert.assertEquals(nodeSchema, nodeSchema.getField("next").getSchema());
+
+    // Serialize and deserialize and validate
+    Assert.assertEquals(nodeSchema, Schema.parseJson(nodeSchema.toString()));
+    Assert.assertEquals(nodeSchema, Schema.parseJson(nodeSchema.toString()).getField("next").getSchema());
+
+    // Test multi-levels recursion. It has the following structure:
+    // Parent {
+    //   List<Child> children
+    // }
+    // Child {
+    //   GrantChild child
+    // }
+    // GrantChild {
+    //   Child parent
+    //   Parent grantParent
```

Comment:
nit: you mean `grandChild` and `grandParent`?

Reply:
No. The parent of the `GrantChild` type is `Child`, and the grant parent of the `GrantChild` is `Parent`, as what being constructed in the Schema below